okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,245 @@
+        super.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                if (!backoffStrategy.nextBackoff().isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to complete the request due to an "
+                            + "IOException", exception));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectToNext(request().url());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call, new IOException("Failed to determine valid failover URL"
+                                + "for '" + request().url() + "' and base URLs " + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                .enqueue(callback);
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                // Relay successful responses
+                if (response.code() / 100 == 2) {
+                    callback.onResponse(call, response);
+                }
+
+                // Handle to handle QoS situations: retry, failover, etc.
+                Optional<QosException> qosError = qosHandler.handle(response);
+                if (qosError.isPresent()) {
+                    qosError.get().accept(createQosVisitor(callback, call));
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
+                if (ioException.isPresent()) {
+                    callback.onFailure(call, ioException.get());
+                    return;
+                }
+
+                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+            }
+        });
+    }
+
+    private QosException.Visitor<Void> createQosVisitor(Callback callback, Call call) {
+        return new QosException.Visitor<Void>() {
+            @Override
+            public Void visit(QosException.Throttle exception) {
+                Optional<Duration> backoff = exception.getRetryAfter().isPresent()
```

Comment:
Do we want to give the backoff strategy a hint (the retry-after header), but ultimately allow it to decide what the backoff should be? Currently a service that constantly returns retry-after will cause us to retry indefinitely

Reply:
good call. added behavior and tests.